File: test-projects/go-stdlib/_seeds/solutions/go-full-moderate/discount.go
package store

// DiscountCode is a stacking discount. Kind is either "percent" (Value is a
// percentage off the current running total) or "fixed" (Value is a flat number
// of cents to subtract).
type DiscountCode struct {
	Kind  string
	Value int
}

// ApplyCodes applies codes IN ORDER to a running total.
//   - "percent": takes Value percent off the current total. A percent above 100
//     is clamped to 100 (and never negative). The new total is rounded to the
//     nearest cent.
//   - "fixed": subtracts Value cents.
//
// The total never drops below 0.
func ApplyCodes(subtotalCents int, codes []DiscountCode) int {
	total := subtotalCents
	for _, code := range codes {
		switch code.Kind {
		case "percent":
			p := code.Value
			if p < 0 {
				p = 0
			}
			if p > 100 {
				p = 100
			}
			remaining := 100 - p
			// Round to the nearest cent: (total*remaining + 50) / 100.
			total = (total*remaining + 50) / 100
		case "fixed":
			total -= code.Value
		}
		if total < 0 {
			total = 0
		}
	}
	return total
}

// CartTotalAfterCodes applies discount codes to the cart subtotal.
func CartTotalAfterCodes(items []LineItem, codes []DiscountCode) int {
	return ApplyCodes(CartSubtotalCents(items), codes)
}

File: test-projects/python-pytest/_seeds/solutions/py-full-moderate/discount.py
"""Stacking discount codes.

A code is a dict: {"kind": "percent", "value": <pct>} or {"kind": "fixed", "value": <cents>}.
Codes apply IN ORDER to a running total. Percent codes take that percent off the current total
(a percent above 100 is clamped to 100, a negative percent is clamped to 0; the resulting total is
rounded to the nearest cent). Fixed codes subtract a number of cents. The total never drops below 0.
"""

from cart import cart_subtotal_cents


def apply_codes(subtotal_cents: int, codes) -> int:
    """Apply discount codes in order to a running total, flooring at 0 after each step."""
    total = subtotal_cents
    for code in codes:
        kind = code["kind"]
        if kind == "percent":
            pct = max(0, min(code["value"], 100))
            total = round(total * (100 - pct) / 100)
        elif kind == "fixed":
            total = total - code["value"]
        else:
            raise ValueError(f"unknown discount kind: {kind!r}")
        if total < 0:
            total = 0
    return total


def cart_total_after_codes(items, codes) -> int:
    """Apply discount codes to a cart's subtotal."""
    return apply_codes(cart_subtotal_cents(items), codes)
